### PR TITLE
Add support for Python binding for disasm_iter

### LINKED ISSUE #2131: Support cs_disasm_iter in Python API binding
Great project ty!
I am disassembling a large binary file and noticing that the slowdown is in cs_disasm (even using _lite version)

Would be great if the Python bindings supported cs_disasm_iter to cut down on memory allocations.

### PR DESCRIPTION
**Pull request by fzakaria**
Add support for `disasm_iter` in the `Cs` class.

Note: I am clear on the performance of this implementation. While the C-implementation
makes claims how it's faster due to less memory allocations, my unfamiliarity with Python
does not give me the same strong assurance.

This provides merely the implementation and hopefully others can do benchmarking.

Here you can see it producing the same values between `iter` and `lite`:
```
❯ python test_lite.py | tail
0x100b:	pshs	cc, b, x, u
0x100d:	lda	, x++
0x100f:	sta	32767, x
0x1013:	lda	[$2017, pcr]
0x1017:	sta	[, x++]
0x1019:	lda	[$1000]
0x101d:	cmps	[4096, x]
0x1022:	rts	
0x1023:

❯ python test_iter.py | tail
0x100b:	pshs	cc, b, x, u
0x100d:	lda	, x++
0x100f:	sta	32767, x
0x1013:	lda	[$2017, pcr]
0x1017:	sta	[, x++]
0x1019:	lda	[$1000]
0x101d:	cmps	[4096, x]
0x1022:	rts	
0x1023:
```

fixes #2131

### CODE REVIEW

**Comment by fzakaria** [comment]
CC @kabeor

File: bindings/python/Makefile [hunk 1]
```diff
@@ -72,6 +72,7 @@ TESTS = test_basic.py test_detail.py test_arm.py test_arm64.py test_m68k.py test
 TESTS += test_ppc.py test_sparc.py test_systemz.py test_x86.py test_xcore.py test_tms320c64x.py
 TESTS += test_m680x.py test_skipdata.py test_mos65xx.py test_bpf.py test_riscv.py
 TESTS += test_evm.py
+TESTS += test_lite.py test_iter.py
```
Comment (fzakaria) [inline bindings/python/Makefile:75]: I actually noticed that test_lite.py was missing also.

File: bindings/python/capstone/__init__.py [hunk 2]
```diff
@@ -511,6 +511,8 @@ def _setup_prototype(lib, fname, restype, *argtypes):
 _setup_prototype(_cs, "cs_open", ctypes.c_int, ctypes.c_uint, ctypes.c_uint, ctypes.POINTER(ctypes.c_size_t))
 _setup_prototype(_cs, "cs_disasm", ctypes.c_size_t, ctypes.c_size_t, ctypes.POINTER(ctypes.c_char), ctypes.c_size_t, \
         ctypes.c_uint64, ctypes.c_size_t, ctypes.POINTER(ctypes.POINTER(_cs_insn)))
+_setup_prototype(_cs, "cs_disasm_iter", ctypes.c_bool, ctypes.c_size_t, ctypes.POINTER(ctypes.POINTER(ctypes.c_char)), ctypes.POINTER(ctypes.c_size_t), \
```
Comment (fzakaria) [inline bindings/python/capstone/__init__.py:514]: I chose to copy ctypes.c_char for the argtype like cs_disasm but the C API uses uint8_t which should map to c_ubyte.

File: bindings/python/capstone/__init__.py [hunk 3]
```diff
@@ -1221,6 +1222,32 @@ def disasm(self, code, offset, count=0):
...
+        insn = _cs_insn()
+        size = ctypes.c_size_t(len(code))
+
+        # Pass a bytearray by reference
+        view = memoryview(code)
+        code = ctypes.pointer(ctypes.c_char.from_buffer_copy(view))
+        if view.readonly:
+            code = (ctypes.c_char * len(view)).from_buffer_copy(view)
+        else: 
+            code = ctypes.pointer(ctypes.c_char.from_buffer(view))
```
Comment (fzakaria) [inline bindings/python/capstone/__init__.py:1243]: Slightly different here than the disasm_lite implementation.
I think it's the same effect of "copy if I can't write"

File: bindings/python/test_iter.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+PPC_CODE = b"\x80\x20\x00\x00\x80\x3f\x00\x00\x10\x43\x23\x0e\xd0\x44\x00\x80\x4c\x43\x22\x02\x2d\x03\x00\x80\x7c\x43\x20\x14\x7c\x43\x20\x93\x4f\x20\x00\x21\x4c\xc8\x00\x21"
+PPC_CODE2 = b"\x10\x60\x2a\x10\x10\x64\x28\x88\x7c\x4a\x5d\x0f"
+SPARC_CODE = b"\x80\xa0\x40\x02\x85\xc2\x60\x08\x85\xe8\x20\x01\x81\xe8\x00\x00\x90\x10\x20\x01\xd5\xf6\x10\x16\x21\x00\x00\x0a\x86\x00\x40\x02\x01\x00\x00\x00\x12\xbf\xff\xff\x10\xbf\xff\xff\xa0\x02\x00\x09\x0d\xbf\xff\xff\xd4\x20\x60\x00\xd4\x4e\x00\x16\x2a\xc2\x80\x03"
+SPARCV9_CODE = b"\x81\xa8\x0a\x24\x89\xa0\x10\x20\x89\xa0\x1a\x60\x89\xa0\x00\xe0"
+SYSZ_CODE = b"\xed\x00\x00\x00\x00\x1a\x5a\x0f\x1f\xff\xc2\x09\x80\x00\x00\x00\x07\xf7\xeb\x2a\xff\xff\x7f\x57\xe3\x01\xff\xff\x7f\x57\xeb\x00\xf0\x00\x00\x24\xb2\x4f\x00\x78"
+XCORE_CODE = b"\xfe\x0f\xfe\x17\x13\x17\xc6\xfe\xec\x17\x97\xf8\xec\x4f\x1f\xfd\xec\x37\x07\xf2\x45\x5b\xf9\xfa\x02\x06\x1b\x10"
+M68K_CODE = b"\xd4\x40\x87\x5a\x4e\x71\x02\xb4\xc0\xde\xc0\xde\x5c\x00\x1d\x80\x71\x12\x01\x23\xf2\x3c\x44\x22\x40\x49\x0e\x56\x54\xc5\xf2\x3c\x44\x00\x44\x7a\x00\x00\xf2\x00\x0a\x28\x4E\xB9\x00\x00\x00\x12\x4E\x75"
```
Comment (fzakaria) [inline bindings/python/test_iter.py:29]: I wish we captured the expected disassembly here :)

Maybe a follow up issue or PR could be to assert the values are correct and adopt pytest?

**Comment by kabeor** [comment]
LGTM, Thanks! Merged.

File: bindings/python/capstone/__init__.py [hunk 5]
```diff
@@ -1221,6 +1222,32 @@ def disasm(self, code, offset, count=0):
             return
             yield
 
+    # This function matches the cs_disasm_iter implementation which
```
Comment (tanclary) [inline bindings/python/capstone/__init__.py:1225]: what's the difference between disasm and diasm (from the PR title)? Or just a typo

Comment (fzakaria) [reply bindings/python/capstone/__init__.py:1225]: typo fixed in title.

**Pull request merged**